homeassistant/components/mqtt/debug_info.py

```diff
@@ -0,0 +1,166 @@
+"""Helper to handle a set of topics to subscribe to."""
+import asyncio
+from functools import partial, wraps
+import logging
+from typing import Any
+
+from homeassistant.helpers.typing import HomeAssistantType
+
+from .const import ATTR_DISCOVERY_PAYLOAD, ATTR_DISCOVERY_TOPIC
+from .models import MessageCallbackType
+
+_LOGGER = logging.getLogger(__name__)
+
+DATA_MQTT_DEBUG_INFO = "mqtt_debug_info"
+STORED_MESSAGES = 10
+
+
+def log_messages(hass: HomeAssistantType, entity_id: str) -> MessageCallbackType:
+    """Wrap an MQTT message callback to support message logging."""
+
+    def _log_message(msg):
+        """Log message."""
+        debug_info = hass.data[DATA_MQTT_DEBUG_INFO]
+        messages = debug_info["entities"][entity_id]["topics"][msg.topic]
+        if len(messages) >= STORED_MESSAGES:
+            messages.pop(0)
+        messages.append(msg.payload)
+
+    def _decorator(msg_callback: MessageCallbackType):
+        # Check for partials to properly determine if coroutine function
+        check_func = msg_callback
+        while isinstance(check_func, partial):
+            check_func = check_func.func
+
+        wrapper_func = None
+        if asyncio.iscoroutinefunction(check_func):
```

Comment:
Is this necessary, are there any that are coroutines ?

Reply:
No, it's not really necessary, non of the native MQTT integrations have coroutine callbacks.
Should I just remove this case?

Reply:
Yeah let's remove it.